### PR TITLE
[NEW] new module 'CashBox Opening from Backend'

### PR DESCRIPTION
**Pull request by GabbasovDinar**

### CODE REVIEW

**Comment by GabbasovDinar** [comment]
@yelizariev check it please

File: pos_cashbox/static/src/js/cashbox.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+    var _t = core._t;
+
+
+    WidgetButton.include({
+        on_click: function(){
+            var self = this;
+            if (this.node.attrs.special === 'open_backend_cashbox'){
+                var config_id = this.view.datarecord.config_id[0];
+                new Model('pos.config').call("search_read", [[['id', '=', config_id]], ["proxy_ip"]]).then(function(res) {
```
Comment (yelizariev) [inline pos_cashbox/static/src/js/cashbox.js]: Try to get proxy_ip from the form directly (add invisible field if needed)

File: pos_cashbox/static/src/js/cashbox.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                    var proxy_ip = res[0].proxy_ip;
+                    if (!proxy_ip) {
+                        return self.show_warning_message(_t('Connection Refused. Please check the connection to PosBox'));
```
Comment (yelizariev) [inline pos_cashbox/static/src/js/cashbox.js]: Fix warning message

File: pos_cashbox/static/src/js/tour.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        position: 'bottom',
+    }, {
+        trigger: ".o_pos_kanban button.oe_kanban_action_button",
+        content: "<p>Click to start the point of sale interface. It <b>runs on tablets</b>, laptops, or industrial hardware.</p><p>Once the session launched, the system continues to run without an internet connection.</p>",
+        position: "bottom"
+    }, {
+        trigger: '.o_main_content:has(.loader:hidden)',
+        content: 'waiting for loading to finish',
+        run: function () {
+            // it's a check
```
Comment (yelizariev) [inline pos_cashbox/static/src/js/tour.js]: Use more clear comment, e.g.

    // specify empty run method to don't click on trigger

File: pos_cashbox/tests/test_default.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+@odoo.tests.common.at_install(True)
+@odoo.tests.common.post_install(True)
+class TestUi(odoo.tests.HttpCase):
+
+    def test_01_cashbox_is_opened(self):
+        self.phantom_js(
+            '/web?debug=assets',
+
+            "odoo.__DEBUG__.services['web_tour.tour']"
+            ".run('pos_cashbox_tour')",
```
Comment (yelizariev) [inline pos_cashbox/tests/test_default.py]: The is not covered your js at all. It opens POS, while your js operates in backend.

Make a tour that

1. opens the widget, clicks the button and gets warning that proxy ip is not specifed
2. via mock js it gets fake response and user sees "wow"

**Comment by flotho** [comment]
Hi @GabbasovDinar , @yelizariev ,

Does this one will solve our Issue? Thanks for your feedback

**Comment by flotho** [comment]
Hi @GabbasovDinar ,

As proposed as a review, the URL you're using is sometimes malformed.
regards

File: pos_cashbox/tests/test_default.py (hunk 4 above)
Comment (GabbasovDinar) [reply pos_cashbox/tests/test_default.py]: @yelizariev We can open a new problem for creating the test

**Pull request merged**